File: src/main/java/com/fouan/actions/MoveSurvivor.java
package com.fouan.actions;

import com.fouan.actors.Survivor;
import com.fouan.actors.view.ActorsView;
import com.fouan.events.ActionChosen;
import com.fouan.game.view.GameView;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
public final class MoveSurvivor implements Action {

    private final GameView gameView;
    private final ActorsView actorsView;

    public MoveSurvivor(GameView gameView, ActorsView actorsView) {
        this.gameView = gameView;
        this.actorsView = actorsView;
    }

    @Override
    public void execute() {
        int turn = gameView.getCurrentTurn();

        actorsView.findCurrentSurvivorForTurn(turn).ifPresent(
            survivor -> survivor.moveTo(null, gameView)
        );
    }

    @EventListener
    public void handleActionChosen(ActionChosen event) {
        if (event.getAction() == Actions.MOVE) {
            execute();
        }
    }
}

File: src/main/java/com/fouan/actions/MoveSurvivor.kt
package com.fouan.actions

import com.fouan.actors.view.ActorsQueries
import com.fouan.events.ActionChosen
import com.fouan.events.AvailableZonesForSurvivorMoveDefined
import com.fouan.events.SurvivorMoved
import com.fouan.events.ZoneChosen
import com.fouan.game.view.GameView
import com.fouan.zones.view.ZonesQueries
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

@Component
class MoveSurvivor(
    private val gameView: GameView,
    private val actorsQueries: ActorsQueries,
    private val zonesQueries: ZonesQueries
) : SurvivorAction {

    @EventListener
    fun handleActionChosen(event: ActionChosen) {
        if (event.action !== Actions.MOVE) return

        val zones = actorsQueries.findCurrentSurvivorIdForTurn(event.turn)
                .flatMap { zonesQueries.findByActorId(it) }
                .map { zonesQueries.findConnectedZones(it) }
                .orElseThrow()
        gameView.fireEvent(AvailableZonesForSurvivorMoveDefined(event.turn, zones))
    }

    @EventListener
    fun handleZoneChosen(event: ZoneChosen) {
        actorsQueries.findCurrentSurvivorIdForTurn(event.turn)
                .flatMap { actorsQueries.findLivingSurvivorBy(it) }
                .ifPresent { gameView.fireEvent(SurvivorMoved(event.turn, it.id, event.position)) }
    }

    override fun getAction(): Actions {
        return Actions.MOVE
    }

    override fun isPossible(): Boolean {
        return actorsQueries.findCurrentSurvivorIdForTurn(gameView.currentTurn)
            .flatMap { zonesQueries.findByActorId(it) }
            .map { zonesQueries.findConnectedZones(it) }
            .orElseThrow()
            .isNotEmpty()
    }
}